openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/EntityRepository.java

```diff
@@ -1421,41 +1425,100 @@ private void validateExtension(T entity) {
+            customPropertyType,
+            propertyConfig);
       } catch (DateTimeParseException e) {
         throw new IllegalArgumentException(
-            CatalogExceptionMessage.dateTimeValidationError(
-                fieldName, TypeRegistry.getCustomPropertyConfig(entityType, fieldName)));
+            CatalogExceptionMessage.dateTimeValidationError(fieldName, propertyConfig));
       }
-      Set<ValidationMessage> validationMessages = jsonSchema.validate(fieldValue);
+
+      Set<ValidationMessage> validationMessages = jsonSchema.validate(entry.getValue());
       if (!validationMessages.isEmpty()) {
         throw new IllegalArgumentException(
             CatalogExceptionMessage.jsonValidationError(fieldName, validationMessages.toString()));
       }
     }
   }
 
+  private void validateAndUpdateExtensionBasedOnPropertyType(
+      T entity,
+      ObjectNode jsonNode,
+      String fieldName,
+      JsonNode fieldValue,
+      String customPropertyType,
+      String propertyConfig) {
+    DateTimeFormatter formatter;
+    switch (customPropertyType) {
+      case "date-cp":
+        DateTimeFormatter inputFormatter =
+            DateTimeFormatter.ofPattern(propertyConfig, Locale.ENGLISH);
+
+        // Parse the input string into a TemporalAccessor
+        TemporalAccessor date = inputFormatter.parse(fieldValue.textValue());
+
+        // Create a formatter for the desired output format
+        DateTimeFormatter outputFormatter =
+            DateTimeFormatter.ofPattern(propertyConfig, Locale.ENGLISH);
+        jsonNode.put(fieldName, outputFormatter.format(date));
+        break;
+      case "dateTime-cp":
+        formatter = DateTimeFormatter.ofPattern(propertyConfig);
+        LocalDateTime dateTime = LocalDateTime.parse(fieldValue.textValue(), formatter);
+        jsonNode.put(fieldName, dateTime.format(formatter));
+        break;
+      case "time-cp":
+        formatter = DateTimeFormatter.ofPattern(propertyConfig);
+        LocalTime time = LocalTime.parse(fieldValue.textValue(), formatter);
+        jsonNode.put(fieldName, time.format(formatter));
+        break;
+      case "metaEnum":
+        JsonNode propertyConfigNode = JsonUtils.readTree(propertyConfig);
+        MetaEnumConfig config = JsonUtils.treeToValue(propertyConfigNode, MetaEnumConfig.class);
+
+        if (!config.getMultiSelect() && fieldValue.size() > 1) {
+          throw new IllegalArgumentException(
+              "Only one key is allowed for non-multiSelect metaEnum");
+        }
+        // Replace each metaEnum key in the fieldValue with the corresponding object from the
+        // propertyConfig
+        Map<String, JsonNode> keyToObjectMap =
+            StreamSupport.stream(propertyConfigNode.get("values").spliterator(), false)
+                .collect(Collectors.toMap(node -> node.get("key").asText(), node -> node));
+
+        if (fieldValue.isArray()) {
+          ArrayNode newArray = JsonNodeFactory.instance.arrayNode();
+
```

Comment:
we can use `JsonUtils` functions here as well I believe.
`ArrayNode newArrayNode = JsonUtils.getObjectNode().arrayNode();`

Reply:
thanks @Siddhanttimeline for the suggestion, have added the change